File: 0744_nextGreatestLetter.py
class Solution:
    def nextGreatestLetter(self, letters: List[str], target: str) -> str:
        if target < letters[0] or target >= letters[-1]: return letters[0]
        for l in letters:
            if target < l:
                return l

# Alternative solution


class Solution:
    def nextGreatestLetter(self, letters, target):
        left, right = 0, len(letters) - 1
        while left <= right:
            mid = left + (right - left) // 2
            if letters[mid] <= target:
                left = mid + 1
            else:
                right = mid - 1
        return letters[(right + 1) % len(letters)]

File: 0501-1000/0744_nextGreatestLetter.java
class Solution {
    public char nextGreatestLetter(char[] letters, char target) {
        int low = 0;
        int high = letters.length - 1;
        int mid;

        while (low <= high) {
            mid = low + (high - low) / 2;

            if (letters[mid] <= target) low = mid + 1;
            else high = mid - 1;
        }

        if (low == letters.length) return letters[0];
        else if (high == -1) return letters[0];
        else return letters[low];
    }
}
